python/plugins/processing/gui/AlgorithmExecutor.py

```diff
@@ -93,13 +91,46 @@ def make_features_compatible(new_features, input_layer, context):
     result_features = []
     for new_f in new_features:
         # Fix attributes
-        if len(new_f.attributes()) > len(input_layer.fields()):
+        if new_f.fields().count() > 0:
+            attributes = []
+            for field in input_layer.fields():
+                if new_f.fields().indexFromName(field.name()) >= 0:
+                    attributes.append(new_f[field.name()])
+                else:
+                    attributes.append(None)
             f = QgsFeature(input_layer.fields())
+            f.setAttributes(attributes)
             f.setGeometry(new_f.geometry())
-            f.setAttributes(new_f.attributes()[:len(input_layer.fields())])
             new_f = f
-        # Fix geometry
-        if new_f.geometry().wkbType() != input_wkb_type:
+        else:
+            lendiff = len(new_f.attributes()) - len(input_layer.fields())
+            if lendiff > 0:
+                f = QgsFeature(input_layer.fields())
+                f.setGeometry(new_f.geometry())
+                f.setAttributes(new_f.attributes()[:len(input_layer.fields())])
+                new_f = f
+            elif lendiff < 0:
+                f = QgsFeature(input_layer.fields())
+                f.setGeometry(new_f.geometry())
+                attributes = new_f.attributes() + [None for i in range(-lendiff)]
+                f.setAttributes(attributes)
+                new_f = f
+
+        # Check if we need geometry manipulation
+        new_f_geom_type = QgsWkbTypes.geometryType(new_f.geometry().wkbType())
+        new_f_has_geom = new_f_geom_type not in (QgsWkbTypes.UnknownGeometry, QgsWkbTypes.NullGeometry)
+        input_layer_has_geom = input_wkb_type not in (QgsWkbTypes.NoGeometry, QgsWkbTypes.Unknown)
+        if input_layer_has_geom and not new_f_has_geom:
+            continue  # Skip this feature completely
```

Comment:
this should just append to result_features -- it's allowable to remove geometries from features and save to a layer with geometry